### PR TITLE
Userland make fixes

### PR DESCRIPTION
**Pull request by ppannuto**
Not yet tested on any HW (will do tomorrow), but all apps build correctly.

@bradjc this should just about earn that damn cookie.

### CODE REVIEW

**Comment by alevy** [comment]
Uh oh, we conflicted: #144

**Comment by ppannuto** [comment]
@alevy The conflict was trivial. Rebased onto #144.

**Comment by alevy** [comment]
@ppannuto Oh no... I was actually just confused

**Comment by alevy** [comment]
I'm happy merging this if #144 makes @bradjc (or anyone else really) happy too.

File: userland/Makefile [hunk 1]
```diff
@@ -19,6 +19,7 @@ CC := $(TOOLCHAIN)-gcc
 CXX := $(TOOLCHAIN)-g++
 # n.b. make convention is that CPPFLAGS are shared for C and C++ sources
 # [CFLAGS is C only, CXXFLAGS is C++ only]
+CFLAGS   += -std=gnu1x
```
Comment (brghena) [inline userland/Makefile]: `gnu1x` is deprecated. Used `gnu11`

File: userland/Makefile [hunk 2]
```diff
@@ -31,25 +32,67 @@ CPPFLAGS += \
 	    -mpic-register=r9\
 	    -mno-pic-data-is-text-relative
 
+$(BUILDDIR)/%.o: %.c | $(BUILDDIR)
+	$(TRACE_DEP)
+	$(Q)$(CC) $(CFLAGS) $(CPPFLAGS) -MF"$(@:.o=.d)" -MG -MM -MP -MT"$(@:.o=.d)@" -MT"$@" "$<"
```
Comment (brghena) [inline userland/Makefile]: Any chance you could add a comment here to explain what this is doing? Why are there all of these arguments here that aren't part of `CFLAGS` or `CPPFLAGS`?

Comment (ppannuto) [inline userland/Makefile]: Fixed

File: userland/Makefile [hunk 3]
```diff
@@ -31,6 +31,9 @@ CPPFLAGS += \
 	    -mpic-register=r9\
 	    -mno-pic-data-is-text-relative
 
+$(BUILDDIR)/%.o: %.c | $(BUILDDIR)
+	$(CC) $(CFLAGS) $(CPPFLAGS) -c -o $@ $<
+
```
Comment (alevy) [inline userland/Makefile]: This was actually intentionally separate. The idea is to separate the responsibilities of the common Makefile and the app specific Makefile.

The app specific Makefile figures out where source files are and how to compile them to object files (e.g. using `gcc` from the main directory, but potentially with another compiler or from other source folders as the case with the ble libraries).

The common Makefile basically plays the role of a platform-specific linker: it's only input is `OBJS` variable (containing object files to be linked), pulls in the right libc and links everything together.

File: userland/Makefile [hunk 4]
```diff
@@ -31,25 +32,69 @@ CPPFLAGS += \
 	    -mpic-register=r9\
 	    -mno-pic-data-is-text-relative
 
+$(BUILDDIR)/%.o: %.c | $(BUILDDIR)
+	# First step doesn't actually compile, just generate header dependency information
+	# More info on our approach here: http://stackoverflow.com/questions/97338
```
Comment (bradjc) [inline userland/Makefile]: Can you add `@`s so these lines don't print?

File: userland/Makefile (hunk 3 above)
Comment (ppannuto) [inline userland/Makefile]: I'm not sold that this is the right choice. That feels like a lot of copy-pasting for a really weird corner case (what app actually needs a different compiler?). It's worth noting that we have some compilation expectations (i.e. `pic`) that people would need to get correct if they rolled their own.

I don't quite understand your ble library example here. The ble-serialization app currently builds with the common ruleset and is actually a great example of how to extend things.

Finally, if people really need to do something different / fancy, they can follow the `libtock` example, just make their own build directory for their fancy stuff and the `$(BUILDDIR)/%.o` pattern won't match. That feels like a fine escape hatch for the esoteric case while making the common case easy.

Comment (bradjc) [inline userland/Makefile]: Yeah example apps should be easy mode. You can always write your own Makefile.

**Comment by alevy** [comment]
@bradjc why is this blocked?

**Comment by bradjc** [comment]
I was holding out for a single app makefile in userland/examples that was then included in each app.

**Comment by ppannuto** [comment]
It'll come, but I'd rather merge this now and then keep improving.

**Comment by alevy** [comment]
I'd rather merge as is as well. I'm actually pretty against a single Makefile for the same reason as the factored out build rule, but it would be even more extreme. The examples should make figuring out how to write an app out-of-repo easy. The fewer levels of indirection the better, there are already basically no rules in the app-specific Makefiles.

**Pull request merged**